pkg/controller/cronjob/utils.go

```diff
@@ -108,43 +107,13 @@ func getRecentUnmetScheduleTimes(sj batchv1beta1.CronJob, now time.Time) ([]time
 		// CronJob as last known start time.
 		earliestTime = sj.ObjectMeta.CreationTimestamp.Time
 	}
-	if sj.Spec.StartingDeadlineSeconds != nil {
-		// Controller is not going to schedule anything below this point
-		schedulingDeadline := now.Add(-time.Second * time.Duration(*sj.Spec.StartingDeadlineSeconds))
-
-		if schedulingDeadline.After(earliestTime) {
-			earliestTime = schedulingDeadline
-		}
-	}
 	if earliestTime.After(now) {
-		return []time.Time{}, nil
+		return nil, nil
 	}
-
-	for t := sched.Next(earliestTime); !t.After(now); t = sched.Next(t) {
-		starts = append(starts, t)
-		// An object might miss several starts. For example, if
-		// controller gets wedged on friday at 5:01pm when everyone has
-		// gone home, and someone comes in on tuesday AM and discovers
-		// the problem and restarts the controller, then all the hourly
-		// jobs, more than 80 of them for one hourly scheduledJob, should
-		// all start running with no further intervention (if the scheduledJob
-		// allows concurrency and late starts).
-		//
-		// However, if there is a bug somewhere, or incorrect clock
-		// on controller's server or apiservers (for setting creationTimestamp)
-		// then there could be so many missed start times (it could be off
-		// by decades or more), that it would eat up all the CPU and memory
-		// of this controller. In that case, we want to not try to list
-		// all the missed start times.
-		//
-		// I've somewhat arbitrarily picked 100, as more than 80,
-		// but less than "lots".
-		if len(starts) > 100 {
```

Comment:
Have you benchmarked the worst-case scenarios for removing this?

Given this specific code, the limit exists for a reason. The cost is small but nontrivial to iterate over a long time window (say, a minutely cron that has failed for a year), and it really adds up with the current synchronous CronJob processing.

There are ways to avoid traversing the entire time window, and thereby avoiding the need for this cap: https://github.com/kubernetes/enhancements/pull/978#discussion_r393457336
I was planning on open-sourcing a stopgap solution (example 2 in that comment),
until a better solution is implimented in the CronJob GA work.

cc @yangkev

Reply:
Thanks for your comment!  This PR actually removes the entire `for` loop, i.e. there is no iteration to be limited any more. The split diff makes this a lot easier to see than the default unified diff.